beluga-aws-sdk/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+use beluga_mqtt::{MqttClient, QoS, Subscriber};
+use beluga_tunnel::Tunnel;
+use error::Error;
+use tokio::task::JoinHandle;
+
+mod error;
+
+type Result<T> = std::result::Result<T, Error>;
+
+pub struct TunnelManager {
+    subscriber: Subscriber,
+}
+
+impl TunnelManager {
+    /// Creates a new [`TunnelManager`] instance by subscribing to the specified
+    /// MQTT topic for tunnel notifications.
+    ///
+    /// This function creates a new [`TunnelManager`] instance by subscribing to
+    /// the MQTT topic `$aws/things/{thing_name}/tunnels/notify` with a quality
+    /// of service level of at least once. The resulting
+    /// [`MqttSubscriber`](Subscriber) is stored in the [`TunnelManager`]
+    /// instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `mqtt` - The `MqttClient` instance to use for the MQTT subscription.
+    /// * `thing_name` - The name of the AWS IoT thing to subscribe to.
+    pub async fn new(mqtt: MqttClient, thing_name: &str) -> Result<Self> {
+        let subscriber = mqtt
+            .subscribe(
+                format!("$aws/things/{thing_name}/tunnels/notify").as_str(),
+                QoS::AtLeastOnce,
+            )
+            .await?;
+
+        Ok(Self { subscriber })
+    }
+
+    /// Receives a packet from the subscriber and spawns a new task to start a
+    /// tunnel.
+    ///
+    /// This function is responsible for handling the incoming packet from the
+    /// subscriber and creating a new tunnel based on the payload of the
+    /// packet. It then starts the tunnel using the provided `Service`.
+    /// The function returns a `JoinHandle` for the spawned task, which can
+    /// be used to await the completion of the tunnel operation.
+    ///
+    /// # Errors
+    ///
+    /// This function may return an error if there is a problem receiving the
+    /// packet from the subscriber or starting the tunnel.
+    pub async fn recv(&mut self) -> Result<JoinHandle<Result<()>>> {
+        let packet = self.subscriber.recv().await?;
+        Ok(tokio::spawn(async move {
```

Comment:
What do you think about having the `TunnelManager` keep track of the spawned services internally instead of returning a handle to it?

I don't think the caller of `recv()` can do anything useful with the join handle anyway and we need to think of a way to perform a graceful shutdown.

I'm thinking we could use a tokio::task::JoinSet for this. That way we can have a `Drop` impl that calls `abort_all` to make sure we always clean up. For a graceful shutdown, we can add a new method `async fn shutdown(&mut self)` which will send a shutdown signal to the different tunnels (maybe with a [CancellationToken](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html) and then wait for the tasks in the join set to complete.